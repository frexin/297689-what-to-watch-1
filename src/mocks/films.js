const movies = [
  {
    id: 1, name: `Bohemian Rhapsody`, picture: `bohemian-rhapsody.jpg`, genre: `Crime`,
    preview: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`
  },
  {
    id: 2, name: `Aviator`, picture: `aviator.jpg`, genre: `Dramas`,
    preview: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`

  },
  {
    id: 3, name: `What We Do in the Shadows`, picture: `what-we-do-in-the-shadows.jpg`, genre: `Crime`,
    preview: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`
  },
  {
    id: 4, name: `Snatch`, picture: `snatch.jpg`, genre: `Horror`,
    preview: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`
  },
  {
    id: 5, name: `Revenant`, picture: `revenant.jpg`, genre: `Romance`,
    preview: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`
  }
];

export {movies};
