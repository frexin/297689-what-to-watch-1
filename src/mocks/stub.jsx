import React from "react";

const StubComponent = () => {
  return (<div/>);
};

export default StubComponent;
