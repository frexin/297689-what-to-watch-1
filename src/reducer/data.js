import adapter from './../adapter.js';

const initialState = {
  currentGenre: `All genres`,
  currentMovie: null,
  moviesList: [],
  isAuthorizationRequired: false,
  userData: null,
  reviews: [],
  similarMovies: []
};

const ActionType = {
  LOAD_MOVIES: `LOAD_MOVIES`,
  LOAD_REVIEWS: `LOAD_REVIEWS`,
  LOAD_USER: `LOAD_USER`,
  CHANGE_GENRE: `CHANGE_GENRE`,
  CHANGE_AUTH_REQUIRE: `CHANGE_AUTH_REQUIRE`,
  GET_MOVIES_BY_GENRE: `GET_MOVIES_BY_GENRE`,
  GET_MOVIE_BY_ID: `GET_MOVIE_BY_ID`
};

const Operation = {
  loadMovies: () => (dispatch, _getState, api) => {
    return api.get(`/films`)
        .then((resp) => {
          const movies = adapter(resp.data);
          dispatch(ActionCreator.loadMovies(movies));
        });
  },
  auth: (userEmail, userPassword) => (dispatch, _getState, api) => {
    return api.post(`/login`, {email: userEmail, password: userPassword})
        .then((resp) => {
          const userData = adapter(resp.data);

          dispatch(ActionCreator.changeAuthRequire(false));
          dispatch(ActionCreator.loadUser(userData));
        });
  },
  loadMovie: (filmId) => (dispatch, _getState, api) => {
    return api.get(`/comments/${filmId}`)
        .then((resp) => {
          const reviews = resp.data;
          dispatch(ActionCreator.loadReviews([reviews, filmId]));
        });
  },
  checkAuth: () => (dispatch, _getState, api) => {
    return api.get(`/login`)
        .then((resp) => {
          const userData = adapter(resp.data);

          dispatch(ActionCreator.changeAuthRequire(false));
          dispatch(ActionCreator.loadUser(userData));
        }).catch(() => {
          dispatch(ActionCreator.changeAuthRequire(true));
        });
  },
};

const reducer = (state = initialState, action) => {
  switch (action.type) {
    case ActionType.LOAD_MOVIES:
      return Object.assign({}, state, {
        moviesList: action.payload
      });

    case ActionType.LOAD_REVIEWS:
      return Object.assign({}, state, {
        reviews: action.payload[0],
        currentMovie: state.moviesList.filter((item) => item.id === action.payload[1])[0]
      });

    case ActionType.CHANGE_GENRE:
      return Object.assign({}, state, {
        currentGenre: action.payload,
      });

    case ActionType.GET_MOVIE_BY_ID:
      return Object.assign({}, state, {
        currentMovie: state.moviesList.filter((item) => item.id === action.payload)[0]
      });

    case ActionType.CHANGE_AUTH_REQUIRE:
      return Object.assign({}, state, {
        isAuthorizationRequired: action.payload
      });

    case ActionType.LOAD_USER:
      return Object.assign({}, state, {
        userData: action.payload
      });
  }

  return state;
};

const ActionCreator = {
  loadMovies: (movies) => {
    return {
      type: ActionType.LOAD_MOVIES,
      payload: movies
    };
  },
  loadReviews: (data) => {
    return {
      type: ActionType.LOAD_REVIEWS,
      payload: data
    };
  },
  changeGenre: (genre) => ({
    type: ActionType.CHANGE_GENRE,
    payload: genre
  }),
  selectMovie: (movieId) => ({
    type: ActionType.GET_MOVIE_BY_ID,
    payload: movieId
  }),
  changeAuthRequire: (authRequire) => ({
    type: ActionType.CHANGE_AUTH_REQUIRE,
    payload: authRequire
  }),
  loadUser: (userData) => ({
    type: ActionType.LOAD_USER,
    payload: userData
  }),
};

export {
  ActionCreator, ActionType, reducer, Operation
};
