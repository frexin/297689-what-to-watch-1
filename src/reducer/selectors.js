import {createSelector} from "reselect";

const getGenreFilter = (state) => state.currentGenre;
const getMovies = (state) => state.moviesList;

export const getMoviesForGenre = createSelector(
    [getGenreFilter, getMovies], (genreFilter, movies) => {
      return movies.filter((movie) => {
        if (genreFilter === `All genres`) {
          return true;
        }

        return movie.genre === genreFilter;
      });
    }
);

export const getGenresList = createSelector(
    [getMovies], (movies) => {
      const genres = [`All genres`];

      movies.forEach((item) => {
        if (!genres.includes(item.genre)) {
          genres.push(item.genre);
        }
      });

      return genres;
    }
);
