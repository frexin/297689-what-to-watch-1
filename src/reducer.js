import {movies} from './mocks/films.js';

const initialState = {
  currentGenre: `All genres`,
  moviesList: []
};

const ActionType = {
  LOAD_MOVIES: `LOAD_MOVIES`,
  CHANGE_GENRE: `CHANGE_GENRE`,
  GET_MOVIES_BY_GENRE: `GET_MOVIES_BY_GENRE`
};

const Operation = {
  loadMovies: () => (dispatch, _getState, api) => {
    return api.get('/films')
        .then(resp => {
          dispatch(ActionCreator.loadMovies(resp.data));
        });
  }
};

const reducer = (state = initialState, action) => {
  switch (action.type) {
    case ActionType.LOAD_MOVIES:
      return Object.assign({}, state, {
        moviesList: action.payload
      });

    case ActionType.CHANGE_GENRE:
      return Object.assign({}, state, {
        currentGenre: action.payload,
      });
  }
  return state;
};

const ActionCreator = {
  loadMovies: (movies) => {
    return {
      type: ActionType.LOAD_MOVIES,
      payload: movies
    }
  },
  changeGenre: (genre) => ({
    type: ActionType.CHANGE_GENRE,
    payload: genre
  })
};

export {
  ActionCreator, ActionType, reducer, Operation
};
