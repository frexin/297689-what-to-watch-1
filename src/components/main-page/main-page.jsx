import React, {Fragment} from "react";
import PropTypes from 'prop-types';

import MoviesList from '../movies-list/movies-list.jsx';
import GenresList from "../genres-list/genres-list.jsx";
import {ActionCreator} from "../../reducer/data.js";
import {connect} from "react-redux";
import withActiveItem from "../../hocs/with-active-item";
import {getMoviesForGenre, getGenresList} from "../../reducer/selectors.js";

const WrappedGenres = withActiveItem(GenresList);
const WrappedMovies = withActiveItem(MoviesList);

const MainPage = (props) => {

  return (
    <Fragment>
      <header className="page-header">
        <div className="logo">
          <a className="logo__link">
            <span className="logo__letter logo__letter--1">W</span>
            <span className="logo__letter logo__letter--2">T</span>
            <span className="logo__letter logo__letter--3">W</span>
          </a>
        </div>

        <div className="user-block">{props.userBlock}</div>
      </header>
      <div className={`page-content`}>
        <section className={`catalog`}>
          <WrappedGenres genres={props.genres} activeItem={props.currentGenre} onSelect={props.onGenreSelect} />
          <WrappedMovies movies={props.movies} onLoadMore={props.onLoadMore} hasMoreMovies={props.hasMoreMovies} />
        </section>
      </div>
    </Fragment>
  );
};

MainPage.defaultProps = {
  movies: [],
};

MainPage.propTypes = {
  userBlock: PropTypes.object,
  movies: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.number.isRequired,
    name: PropTypes.string.isRequired,
    previewImage: PropTypes.string.isRequired,
    previewVideoLink: PropTypes.string.isRequired
  })).isRequired,
  onGenreSelect: PropTypes.func.isRequired,
  currentGenre: PropTypes.string.isRequired,
  genres: PropTypes.array.isRequired,
  onLoadMore: PropTypes.func,
  hasMoreMovies: PropTypes.bool
};

const mapStateToProps = (state, ownProps) => Object.assign({}, ownProps, {
  movies: getMoviesForGenre(state),
  genres: getGenresList(state),
  currentGenre: state.currentGenre,
  authRequire: state.isAuthorizationRequired,
  hasMoreMovies: state.hasMoreMovies
});

const mapDispatchToProps = (dispatch) => {
  return {
    onGenreSelect: (genre) => {
      dispatch(ActionCreator.changeGenre(genre));
    },
    onLoadMore: () => {
      dispatch(ActionCreator.extendLimit());
    }
  };
};

export {MainPage};

export default connect(mapStateToProps, mapDispatchToProps)(MainPage);

