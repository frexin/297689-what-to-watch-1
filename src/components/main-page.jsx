import React from "react";
import PropTypes from 'prop-types';

import MoviesList from './movies-list.jsx';
import GenresList from "./genres-list.jsx";
import {ActionCreator} from "../reducer";
import {connect} from "react-redux";
import withActiveItem from "../hocs/with-active-item";
import {getMoviesForGenre, getGenresList} from "../reducer/selectors";

const WrappedGenres = withActiveItem(GenresList);
const WrappedMovies = withActiveItem(MoviesList);

const MainPage = (props) => {

  return (
      <div className={`page-content`}>
        <section className={`catalog`}>
          <WrappedGenres genres={props.genres} activeItem={props.currentGenre}
                         onSelect={props.onGenreSelect}/>
          <WrappedMovies movies={props.movies}/>
        </section>
      </div>
  );
};

MainPage.defaultProps = {
  movies: [],
};

MainPage.propTypes = {
  movies: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.number.isRequired,
    name: PropTypes.string.isRequired,
    preview_image: PropTypes.string.isRequired,
    preview_video_link: PropTypes.string.isRequired
  })).isRequired,
  onGenreSelect: PropTypes.func.isRequired,
  currentGenre: PropTypes.string.isRequired
};

const mapStateToProps = (state, ownProps) => Object.assign({}, ownProps, {
  movies: getMoviesForGenre(state),
  genres: getGenresList(state),
  currentGenre: state.currentGenre
});

const mapDispatchToProps = (dispatch) => {
  return {
    onGenreSelect: (genre) => {
      dispatch(ActionCreator.changeGenre(genre));
    }
  };
};

export {MainPage};

export default connect(mapStateToProps, mapDispatchToProps)(MainPage);

