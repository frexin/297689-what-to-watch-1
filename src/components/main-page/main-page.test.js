import React from 'react';
import renderer from 'react-test-renderer';
import {MainPage} from './main-page.jsx';
import {getGenresList} from '../../reducer/selectors';
import {BrowserRouter} from "react-router-dom";

import {movies} from "../../mocks/films";

const state = {
  currentGenre: `All genres`,
  moviesList: movies,
};

const genres = getGenresList(state);

it(`Main component shows default page`, () => {

  const user = {
    id: 1,
    email: `Oliver.conner@gmail.com`,
    name: `Oliver.conner`
  };

  const tree = renderer.create(<BrowserRouter><MainPage genres={genres} currentGenre={`All genres`} onGenreSelect={jest.fn()} user={user} movies={movies}/></BrowserRouter>).toJSON();

  expect(tree).toMatchSnapshot();
});

it(`Main component shows signin page`, () => {
  const tree = renderer.create(<BrowserRouter><MainPage genres={genres} currentGenre={`All genres`} onGenreSelect={jest.fn()} user={null} movies={movies}/></BrowserRouter>).toJSON();

  expect(tree).toMatchSnapshot();
});
