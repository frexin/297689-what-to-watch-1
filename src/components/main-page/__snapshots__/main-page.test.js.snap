// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main component shows default page 1`] = `
<div
  className="page-content"
>
  <section
    className="catalog"
  >
    <ul
      className="catalog__genres-list"
    >
      <li
        className="catalog__genres-item catalog__genres-item--active"
      >
        <a
          className="catalog__genres-link"
          href="#"
          onClick={[Function]}
        >
          All genres
        </a>
      </li>
      <li
        className="catalog__genres-item "
      >
        <a
          className="catalog__genres-link"
          href="#"
          onClick={[Function]}
        >
          Crime
        </a>
      </li>
      <li
        className="catalog__genres-item "
      >
        <a
          className="catalog__genres-link"
          href="#"
          onClick={[Function]}
        >
          Dramas
        </a>
      </li>
      <li
        className="catalog__genres-item "
      >
        <a
          className="catalog__genres-link"
          href="#"
          onClick={[Function]}
        >
          Horror
        </a>
      </li>
      <li
        className="catalog__genres-item "
      >
        <a
          className="catalog__genres-link"
          href="#"
          onClick={[Function]}
        >
          Romance
        </a>
      </li>
    </ul>
    <div
      className="catalog__movies-list"
    >
      <article
        className="small-movie-card catalog__movies-card"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        <div
          className="small-movie-card__image"
        >
          <img
            alt="Bohemian Rhapsody"
            height={175}
            src="bohemian-rhapsody.jpg"
            width={280}
          />
        </div>
        <h3
          className="small-movie-card__title"
        >
          <a
            className="small-movie-card__link"
            href="/film/1"
            onClick={[Function]}
          >
            Bohemian Rhapsody
          </a>
        </h3>
      </article>
      <article
        className="small-movie-card catalog__movies-card"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        <div
          className="small-movie-card__image"
        >
          <img
            alt="Aviator"
            height={175}
            src="aviator.jpg"
            width={280}
          />
        </div>
        <h3
          className="small-movie-card__title"
        >
          <a
            className="small-movie-card__link"
            href="/film/2"
            onClick={[Function]}
          >
            Aviator
          </a>
        </h3>
      </article>
      <article
        className="small-movie-card catalog__movies-card"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        <div
          className="small-movie-card__image"
        >
          <img
            alt="What We Do in the Shadows"
            height={175}
            src="what-we-do-in-the-shadows.jpg"
            width={280}
          />
        </div>
        <h3
          className="small-movie-card__title"
        >
          <a
            className="small-movie-card__link"
            href="/film/3"
            onClick={[Function]}
          >
            What We Do in the Shadows
          </a>
        </h3>
      </article>
      <article
        className="small-movie-card catalog__movies-card"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        <div
          className="small-movie-card__image"
        >
          <img
            alt="Snatch"
            height={175}
            src="snatch.jpg"
            width={280}
          />
        </div>
        <h3
          className="small-movie-card__title"
        >
          <a
            className="small-movie-card__link"
            href="/film/4"
            onClick={[Function]}
          >
            Snatch
          </a>
        </h3>
      </article>
      <article
        className="small-movie-card catalog__movies-card"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        <div
          className="small-movie-card__image"
        >
          <img
            alt="Revenant"
            height={175}
            src="revenant.jpg"
            width={280}
          />
        </div>
        <h3
          className="small-movie-card__title"
        >
          <a
            className="small-movie-card__link"
            href="/film/5"
            onClick={[Function]}
          >
            Revenant
          </a>
        </h3>
      </article>
    </div>
  </section>
</div>
`;
